### PR TITLE
Update new/edit policy pages

### PR DESCRIPTION
**Pull request by bmanifold**
Why:

* The new and edit policy pages had previously only been pulling live data from the DB, but had not been able to use the forms to create or update policies.  This commit allows the forms to function as intended.

### CODE REVIEW

File: elixir/apps/domain/lib/domain/accounts/account.ex
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/accounts/account.ex]: We should implement this protocol in `web` app, because in `domain` Phoenix is should not be added as a dependency:
```
defimpl Phoenix.Param, for: Domain.Accounts.Account do
  def to_param(account), do: account.slug || account.id
end
```
(web/lib/web/protocols.ex)

Comment (bmanifold) [reply elixir/apps/domain/lib/domain/accounts/account.ex]: Ah, ok, that makes sense. 👍   I'll get this updated

File: elixir/apps/web/lib/web/live/policies/new.ex
Comment (AndrewDryga) [inline elixir/apps/web/lib/web/live/policies/new.ex]: We should have a `new_policy(attrs \\ %{})` function in policies so that `web` app won't have to dig into changesets.

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (AndrewDryga) [inline elixir/apps/web/lib/web/live/policies/index.ex]: This should be an exceptional race condition (somebody deleted while you see it) so crashing here should be fine for now:
```suggestion
  def handle_event("delete", %{"id" => id}, socket) do
      {:ok, _} = Policies.delete_policy(socket.assigns.policy, socket.assigns.subject)
      {:noreply, redirect(socket, to: ~p"/#{socket.assigns.account}/policies")}
  end
```

File: elixir/apps/domain/priv/repo/migrations/20230801154306_create_policies.exs
Comment (jamilbk) [inline elixir/apps/domain/priv/repo/migrations/20230801154306_create_policies.exs]: 👍

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/policies/index.ex]: Out of curiosity, why [`redirect/2`](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#redirect/2) and not [`push_navigate/2`](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#push_navigate/2) here? I believe the former causes a full page reload.

File: elixir/apps/web/lib/web/live/policies/show.ex
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/policies/show.ex]: Same question here Re: `redirect/2` vs `push_navigate/2`.

File: elixir/apps/web/lib/web/live/policies/new.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/new.ex]: Gotcha 👍

**Review by jamilbk — COMMENTED** [review]
LGTM but had a question on full page reloads.

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/index.ex]: Ah, good question/catch.  I think this was again Copilot and since it worked I didn't really think critically about it.  I believe you're correct about the full page reload, so I'll update this.

File: elixir/apps/web/lib/web/live/policies/show.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/show.ex]: Same response as above.  Will get this updated.

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/index.ex]: Sounds good 👍

File: elixir/apps/domain/priv/repo/migrations/20230801154306_create_policies.exs
Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20230801154306_create_policies.exs]: This actually reminded me that I need to pull this out in to it's own migration so we don't have to completely reset the staging DB.

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/index.ex]: After looking at this, don't we have to fetch the policy first here anyway?  We don't have `socket.assigns.policy` for the index page.  We do for the `show` page, so I could update that handler if you want.

File: elixir/apps/domain/lib/domain/accounts/account.ex
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/accounts/account.ex]: Moved this to the `web` app as recommended

File: elixir/apps/web/lib/web/live/policies/new.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/new.ex]: Added the new function as recommended.

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/index.ex]: Updated this to use `push_navigate`

File: elixir/apps/web/lib/web/live/policies/show.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/show.ex]: Updated this to use `push_navigate`

File: elixir/apps/domain/priv/repo/migrations/20230801154306_create_policies.exs
Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20230801154306_create_policies.exs]: Moved this change to it's own migration and reverted this file to it's previous state.

**Comment by jamilbk** [comment]
Rebased to pull in a CI status check fix

**Comment by jamilbk** [comment]
Was this requiring additional approval?

**Comment by bmanifold** [comment]
I think we were going to wait until @AndrewDryga 's PR was merged first if I'm remembering correctly

**Comment by jamilbk** [comment]
> I think we were going to wait until @AndrewDryga 's PR was merged first if I'm remembering correctly

Ah! that's right. Kk.

**Comment by AndrewDryga** [comment]
@bmanifold this one should be ready for rebase and merge.

**Comment by bmanifold** [comment]
👍 Sounds good.  I'll get it rebased and fix any merge conflicts.

File: elixir/apps/domain/lib/domain/policies.ex
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/policies.ex]: There is no need to check permissions here, this will also violate the function return (it will be either a changeset or error tuple, which is not idiomatic),

File: elixir/apps/domain/priv/repo/migrations/20230825174719_update_policy_indexes.exs
Comment (AndrewDryga) [inline elixir/apps/domain/priv/repo/migrations/20230825174719_update_policy_indexes.exs]: ATM we don't support rollbacks:
```suggestion
  def change do
    drop(index(:policies, [:account_id, :name]))
    drop(index(:policies, [:account_id, :resource_id, :actor_group_id]))

    create(index(:policies, [:account_id, :name], unique: true, where: "deleted_at IS NULL"))

    create(
      index(:policies, [:account_id, :resource_id, :actor_group_id],
        unique: true,
        where: "deleted_at IS NULL"
      )
    )
  end
```

File: elixir/apps/domain/lib/domain/policies.ex
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/policies.ex]: 👍 Gotcha.  Since I was having to pass the subject I figured it would be good to go ahead and check permissions, but I didn't pay enough attention to the return type.  Good catch.

File: elixir/apps/domain/priv/repo/migrations/20230825174719_update_policy_indexes.exs
Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20230825174719_update_policy_indexes.exs]: Ah, ok, I didn't realize we weren't supporting rollbacks.  I actually started with a `def change do` 😆

Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20230825174719_update_policy_indexes.exs]: This has been updated

File: elixir/apps/domain/lib/domain/policies.ex
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/policies.ex]: This has been updated

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/index.ex]: I realized this comment was still open, so I went back and checked on what the behavior was both using the `with` and removing it.  Both seemed to crash, so I went ahead and removed the `with` and `else`.  This should be updated now.

File: elixir/apps/web/lib/web/live/policies/edit.ex
Comment (AndrewDryga) [inline elixir/apps/web/lib/web/live/policies/edit.ex]: The form can be temporarily assigned to reduce the state that websocket will hold, since we will rebuild it on every update anyways:
```suggestion
      form = to_form(Policies.Policy.Changeset.update(policy, %{}))
      {:ok, assign(socket, policy: policy), temorary_assigns: [form: form]}
```

File: elixir/apps/web/lib/web/live/policies/index.ex
Comment (AndrewDryga) [inline elixir/apps/web/lib/web/live/policies/index.ex:100]:
```suggestion
    {:ok, _} = Policies.delete_policy(socket.assigns.policy, socket.assigns.subject)
```

File: elixir/apps/web/lib/web/live/policies/new.ex
Comment (AndrewDryga) [inline elixir/apps/web/lib/web/live/policies/new.ex]: I think this class should not have an effect because the `.simple_form` component already wraps everything in `<div class="space-y-8 bg-white">`. Maybe we should drop this component completely? Doesn't look like it has a lot of usage anyway but it makes code less consistent.

File: elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs
Comment (AndrewDryga) [inline elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs:77]: Here, we should still use the ID to redirect from IdP to Firezone. We don't want the sync to break when the slug is changed while those URLs must be always up-to-date and whitelisted on the IdP end.

**Comment by AndrewDryga** [comment]
A few small things and LGTM. Also, PR needs a rebase.

File: elixir/apps/web/lib/web/live/policies/edit.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/edit.ex]: 👍 makes sense.  I've update this and the Policy `new.ex` liveview with the temporary assigns.

Also, I tried the suggested code as is, but it did not work.  It looks like you need to assign it to the socket using the `assign` function and then in the `temporary_assigns` value you tell it which assigns are temporary and what the value of the assign should be reset to.

File: elixir/apps/web/lib/web/live/policies/new.ex
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/policies/new.ex]: Yeah, I think you're right about the class not having any effect.

As far as getting rid of the component, I don't have a strong preference.  I started using the `.simple_form` component in the `sign_up.ex` page because I saw it was used in the `auth/sign_in.ex` page and then I used it here because I used it in the `sign_up.ex` page.

I think my only suggestion would be to leave it for now and then have another PR that goes through and removes it from everything.  That way this PR can stay focused to just the Policy New/Edit pages and then the new PR can focus on removing the component and all of it's uses.  Thoughts?

File: elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs
Comment (bmanifold) [reply elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs:77]: This is a good point.  This means I'll need to go in and update the code that does the redirect to the IDP to only use the `account.id` rather than accepting an `account` and letting the protocol handle the interpolation.  

I'll also need to make sure that what's displayed to the user for setting up their IDP is shown as a URL with the `account.id` rather than a URL with the `account.slug`.

In keeping with my previous comment, how you would feel about leaving this as is for now and me putting up a more targeted follow on PR for this?

Comment (AndrewDryga) [reply elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs:77]: I think the problem here is that this PR changes it to purely use slugs in the first place so it feels not right to introduce an issue in code not related to the PR to fix it later in another one. 

Maybe let's move it out to a separate PR which will enforce slug usage everywhere except with IdP redirects? Should be pretty easy to split the code.

Comment (bmanifold) [reply elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs:77]: Ok, that's fair.  Yeah, I can remove the code to default the URLs to use slugs and then open a new PR for just that.

**Review by AndrewDryga — APPROVED** [review]

File: elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs
Comment (bmanifold) [reply elixir/apps/web/test/web/live/settings/identity_providers/google_workspace/connect_test.exs:77]: I've removed the code that defaulted all references of account to `account.slug`.  Also updated/reverted any changed tests.

**Pull request merged**
